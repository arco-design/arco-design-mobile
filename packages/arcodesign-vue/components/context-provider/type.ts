export interface ContextInstallOptions {
    /**
     * 组件名称前缀
     * @en Component prefix
     * @default "Arco"
     */
    componentPrefix?: string;
}

export interface GlobalContextParams {
    /**
     * 组件类名前缀
     * @en Component classname prefix
     * @default "arco"
     */
    prefixCls?: string;
}
