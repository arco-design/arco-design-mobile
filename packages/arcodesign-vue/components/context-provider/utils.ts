import { InjectionKey, inject } from 'vue';
import { ContextInstallOptions, GlobalContextParams } from './type';

export const defaultInstallOptions: Required<ContextInstallOptions> = {
    componentPrefix: 'Arco',
};

export const defaultContext: Required<GlobalContextParams> = {
    prefixCls: 'arco',
};

export const configProviderInjectionKey: InjectionKey<GlobalContextParams> =
    Symbol('ArcoContextProvider');

export function getComponentPrefix(options?: ContextInstallOptions) {
    return options?.componentPrefix || defaultInstallOptions.componentPrefix;
}

export function getPrefixCls(componentName?: string) {
    const contextProvider = inject(configProviderInjectionKey, defaultContext);
    const prefix = contextProvider.prefixCls;
    if (componentName) {
        return `${prefix}-${componentName}`;
    }
    return prefix;
}
