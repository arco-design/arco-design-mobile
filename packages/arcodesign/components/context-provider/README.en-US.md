### Others

# ContextProvider 

Global data control component, used to replace global data.

======

> Props

|Property|Description|Type|DefaultValue|
|----------|-------------|------|------|
|prefixCls|Component classname prefix|string|"arco"|
|system|Manually control the current system, and the incoming value will be used directly after being passed in\. It is applicable when the initial value of the system needs to be specified in the ssr scenario\.|"" \| "pc" \| "android" \| "ios"|""|
|useDarkMode|Whether to use dark mode|boolean|false|
|isDarkMode|Whether it is in dark mode|boolean|false|
|theme|Theme variable\. The css variable will be replaced online after input\. The less variable needs to be set|Record\<string, string\>|-|
|locale|Internationalized language configuration|ILocale|-|
|useRtl|Whether to use rtl|boolean|false|

> ILocale

|Property|Description|Type|
|----------|-------------|------|
|locale|Language Type|string|
|LoadMore|-|\{ loadMoreText: string; loadingText: string; prepareText: string; noDataText: string; failLoadText: string; prepareScrollText: string; prepareClickText: string; \}|
|Picker|-|\{ okText: string; cancelText: string; \}|
|Tag|-|\{ addTag: string; \}|
|Dialog|-|\{ okText: string; cancelText: string; \}|
|SwipeLoad|-|\{ normalText: string; activeText: string; \}|
|PullRefresh|-|\{ loadingText: string; pullingText: string; finishText: string; loosingText: string; \}|
|DropdownMenu|-|\{ select: string; \}|
|Pagination|-|\{ previousPage: string; nextPage: string; \}|
|Image|-|\{ loadError: string; \}|
|ImagePicker|-|\{ loadError: string; \}|
|SearchBar|-|\{ placeholder: string; cancelBtn: string; \}|
|Stepper|-|\{ minusButtonName: string; addButtonName: string; \}|
|Keyboard|-|\{ confirm: string; \}|
|Form|-|\{ required: string; type: \{ email: string; url: string; string: string; number: string; array: string; object: string; boolean: string; \}; number: \{ min: string; max: string; equal: string; range: string; positive: string; negative: string; \}; string: \{ \.\.\.; \}; array: \{ \.\.\.; \}; object: \{ \.\.\.; \}; boolean: \{ \.\.\.; \}; \}|
