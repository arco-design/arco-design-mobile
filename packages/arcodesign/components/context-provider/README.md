### 其他

# 全局配置 ContextProvider

全局数据控制组件，用于替换全局数据。

======

> 属性/Props

|参数|描述|类型|默认值|
|----------|-------------|------|------|
|prefixCls|组件类名前缀|string|"arco"|
|system|手动控制当前所在系统，传入后将直接使用传入的值，ssr场景需指定系统初始值时适用|"" \| "pc" \| "android" \| "ios"|""|
|useDarkMode|是否使用暗黑模式|boolean|false|
|isDarkMode|是否处于暗黑模式，指定后以指定的值为准|boolean|false|
|theme|主题变量，传入后将在线替换css变量，需设置less变量 @use\-css\-vars: 1|Record\<string, string\>|-|
|locale|国际化语言包配置|ILocale|-|
|useRtl|是否使用Rtl模式|boolean|false|

> ILocale

|参数|描述|类型|
|----------|-------------|------|
|locale|语言类型|string|
|LoadMore|-|\{ loadMoreText: string; loadingText: string; prepareText: string; noDataText: string; failLoadText: string; prepareScrollText: string; prepareClickText: string; \}|
|Picker|-|\{ okText: string; cancelText: string; \}|
|Tag|-|\{ addTag: string; \}|
|Dialog|-|\{ okText: string; cancelText: string; \}|
|SwipeLoad|-|\{ normalText: string; activeText: string; \}|
|PullRefresh|-|\{ loadingText: string; pullingText: string; finishText: string; loosingText: string; \}|
|DropdownMenu|-|\{ select: string; \}|
|Pagination|-|\{ previousPage: string; nextPage: string; \}|
|Image|-|\{ loadError: string; \}|
|ImagePicker|-|\{ loadError: string; \}|
|SearchBar|-|\{ placeholder: string; cancelBtn: string; \}|
|Stepper|-|\{ minusButtonName: string; addButtonName: string; \}|
|Keyboard|-|\{ confirm: string; \}|
|Form|-|\{ required: string; type: \{ email: string; url: string; string: string; number: string; array: string; object: string; boolean: string; \}; number: \{ min: string; max: string; equal: string; range: string; positive: string; negative: string; \}; string: \{ \.\.\.; \}; array: \{ \.\.\.; \}; object: \{ \.\.\.; \}; boolean: \{ \.\.\.; \}; \}|
