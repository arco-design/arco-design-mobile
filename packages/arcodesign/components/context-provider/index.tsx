import React, { createContext, useState, useEffect, useCallback } from 'react';
import { addCssRules, removeCssStyleDom, defaultLocale, ILocale } from '@arco-design/mobile-utils';

export interface GlobalContextParams {
    /**
     * 组件类名前缀
     * @en Component classname prefix
     * @default "arco"
     */
    prefixCls?: string;
    /**
     * 手动控制当前所在系统，传入后将直接使用传入的值，ssr场景需指定系统初始值时适用
     * @en Manually control the current system, and the incoming value will be used directly after being passed in. It is applicable when the initial value of the system needs to be specified in the ssr scenario.
     * @default ""
     */
    system?: 'pc' | 'android' | 'ios' | '';
    /**
     * 是否使用暗黑模式
     * @en Whether to use dark mode
     * @default false
     */
    useDarkMode?: boolean;
    /**
     * 是否处于暗黑模式，指定后以指定的值为准
     * @en Whether it is in dark mode
     * @default false
     */
    isDarkMode?: boolean;
    /**
     * 主题变量，传入后将在线替换css变量，需设置less变量 @use-css-vars: 1
     * @en Theme variable. The css variable will be replaced online after input. The less variable needs to be set @use-css-vars: 1
     */
    theme?: Record<string, string>;
    /**
     * 国际化语言包配置
     * @en Internationalized language configuration
     */
    locale?: ILocale;
    /**
     * 是否使用Rtl模式
     * @en Whether to use rtl
     * @default false
     */
    useRtl?: boolean;
}

export const defaultContext: GlobalContextParams = {
    prefixCls: 'arco',
    system: '',
    useDarkMode: false,
    isDarkMode: false,
    locale: defaultLocale,
    useRtl: false,
};

export const GlobalContext = createContext<GlobalContextParams>(defaultContext);

export interface ContextProviderProps extends GlobalContextParams {
    children: React.ReactNode;
}

export type WithGlobalContext<T> = T & { context?: GlobalContextParams };

/**
 * 全局数据控制组件，用于替换全局数据。
 * @en Global data control component, used to replace global data.
 * @type 其他
 * @type_en Others
 * @name 全局配置
 * @name_en ContextProvider
 */
export default function ContextProvider(props: ContextProviderProps) {
    const {
        children,
        useDarkMode,
        isDarkMode,
        theme,
        locale = defaultLocale,
        ...restProps
    } = props;
    const [isDarkModeState, setIsDarkModeState] = useState<boolean>();
    const changeDarkMode = useCallback(
        (res: MediaQueryListEvent) => setIsDarkModeState(res.matches),
        [],
    );

    useEffect(() => {
        const matchMedia = window.matchMedia('(prefers-color-scheme: dark)');
        if (useDarkMode) {
            const dark = matchMedia.matches;
            setIsDarkModeState(dark);
            if (typeof matchMedia.addEventListener === 'function') {
                matchMedia.addEventListener('change', changeDarkMode);
            } else if (typeof matchMedia.addListener === 'function') {
                matchMedia.addListener(changeDarkMode);
            }
        }
        return () => {
            if (useDarkMode) {
                if (typeof matchMedia.removeEventListener === 'function') {
                    matchMedia.removeEventListener('change', changeDarkMode);
                } else if (typeof matchMedia.removeListener === 'function') {
                    matchMedia.removeListener(changeDarkMode);
                }
            }
        };
    }, [useDarkMode]);

    useEffect(() => {
        if (!theme) {
            return;
        }
        addCssRules('arcoCustomTheme', theme);
        return () => {
            removeCssStyleDom('arcoCustomTheme');
        };
    }, [theme]);

    return (
        <GlobalContext.Provider
            value={{
                ...defaultContext,
                ...restProps,
                isDarkMode: isDarkMode === void 0 ? isDarkModeState : isDarkMode,
                useDarkMode,
                locale: locale || defaultLocale,
            }}
        >
            {children}
        </GlobalContext.Provider>
    );
}

export const ContextLayout = GlobalContext.Consumer;

export function CompWithGlobalContext<P extends JSX.IntrinsicAttributes>(
    Component: React.FunctionComponent<P>,
) {
    return function (props: WithGlobalContext<P>) {
        const { context: propsContext, ...others } = props;
        return (
            <ContextLayout>
                {context => (
                    <ContextProvider {...{ ...context, ...propsContext }}>
                        <Component {...(others as P)} />
                    </ContextProvider>
                )}
            </ContextLayout>
        );
    };
}
