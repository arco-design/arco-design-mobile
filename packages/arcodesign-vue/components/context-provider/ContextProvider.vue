<script lang="ts" setup>
import { provide } from 'vue';
import { GlobalContextParams } from './type';
import { configProviderInjectionKey, defaultContext } from './utils';

const props = withDefaults(defineProps<GlobalContextParams>(), defaultContext);
provide(configProviderInjectionKey, props);
</script>
<template>
    <slot />
</template>
