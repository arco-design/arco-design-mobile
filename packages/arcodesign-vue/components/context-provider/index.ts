import ContextProvider from './ContextProvider.vue';
import { installFactory } from '../_helpers/install';

export type * from './type';
export * from './utils';

const install = installFactory(ContextProvider);

export { ContextProvider };

export default {
    install,
};
