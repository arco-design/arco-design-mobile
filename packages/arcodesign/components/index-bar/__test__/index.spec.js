import React from 'react';
import demoTest from '../../../tests/demoTest';
import mountTest from '../../../tests/mountTest';
import { defaultContext } from '../../context-provider';
import IndexBar from '..';

demoTest('index-bar');

mountTest(IndexBar, 'IndexBar');

const prefix = `${defaultContext.prefixCls}-index-bar`;

describe('IndexBar', () => {

});
