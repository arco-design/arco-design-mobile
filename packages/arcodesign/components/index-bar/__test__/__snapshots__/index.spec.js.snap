// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index-bar demo test index-bar demo: index.md renders correctly 1`] = `
<DocumentFragment>
  <div
    class="arco-tabs arco-tabs-top arco-tabs-vertical all-border-box"
  >
    <div
      class="arco-tab-cell-container-wrap vertical type-line pc"
    >
      <div
        class="arco-tab-cell-container-inner vertical pos-top type-line has-divider pc"
      >
        <div
          class="arco-tab-cell-container vertical pos-top arrange-center type-line shown"
        >
          <div
            class="arco-tab-cell vertical line pc active no-shrink"
          >
            基础样式
          </div>
          <div
            class="arco-tab-cell vertical line pc no-shrink last"
          >
            自定义
          </div>
          <div
            class="arco-tab-cell-underline"
            style="transition-duration: 0ms; transform: translateX(0px) translateZ(0);"
          >
            <div
              class="arco-tab-cell-underline-inner vertical"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="arco-tab-pane-container vertical mode-swipe"
      style="width: 200%; transform: translateX(0px) translateZ(0); transition-duration: 0ms;"
    >
      <div
        class="arco-tab-pane mode-swipe"
      >
        <div
          class="arco-index-bar custom-index-bar-height"
        >
          <div
            class="arco-index-bar-container"
          >
            <div
              class="arco-index-bar-group"
              data-index="A"
            >
              <div>
                <div
                  style="height: 0px;"
                />
                <div
                  class="arco-sticky"
                >
                  <div
                    class="arco-index-bar-group-title arco-index-bar-group-active"
                  >
                    A
                  </div>
                </div>
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                鞍山
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                安庆
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                安阳
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                安康
              </div>
            </div>
            <div
              class="arco-index-bar-group"
              data-index="B"
            >
              <div>
                <div
                  style="height: 0px;"
                />
                <div
                  class="arco-sticky"
                >
                  <div
                    class="arco-index-bar-group-title"
                  >
                    B
                  </div>
                </div>
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                北京
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                保定
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                包头
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                蚌埠
              </div>
            </div>
            <div
              class="arco-index-bar-group"
              data-index="C"
            >
              <div>
                <div
                  style="height: 0px;"
                />
                <div
                  class="arco-sticky"
                >
                  <div
                    class="arco-index-bar-group-title"
                  >
                    C
                  </div>
                </div>
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                成都
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                重庆
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                朝阳
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                长春
              </div>
            </div>
            <div
              class="arco-index-bar-group"
              data-index="D"
            >
              <div>
                <div
                  style="height: 0px;"
                />
                <div
                  class="arco-sticky"
                >
                  <div
                    class="arco-index-bar-group-title"
                  >
                    D
                  </div>
                </div>
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                大成
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                大拟
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                丹东
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                定边
              </div>
            </div>
            <div
              class="arco-index-bar-group"
              data-index="E"
            >
              <div>
                <div
                  style="height: 0px;"
                />
                <div
                  class="arco-sticky"
                >
                  <div
                    class="arco-index-bar-group-title"
                  >
                    E
                  </div>
                </div>
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                鄂尔多斯
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                二道湾
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                二龙
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                峨眉
              </div>
            </div>
            <div
              class="arco-index-bar-group"
              data-index="F"
            >
              <div>
                <div
                  style="height: 0px;"
                />
                <div
                  class="arco-sticky"
                >
                  <div
                    class="arco-index-bar-group-title"
                  >
                    F
                  </div>
                </div>
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                福州
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                奉化
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                福海
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                肥东
              </div>
            </div>
            <div
              class="arco-index-bar-group"
              data-index="G"
            >
              <div>
                <div
                  style="height: 0px;"
                />
                <div
                  class="arco-sticky"
                >
                  <div
                    class="arco-index-bar-group-title"
                  >
                    G
                  </div>
                </div>
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                广州
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                贵阳
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                广安
              </div>
              <div
                class="arco-index-bar-group-item"
              >
                广州南
              </div>
            </div>
          </div>
          <div
            class="arco-index-bar-sidebar"
          >
            <div
              class="arco-index-bar-sidebar-item arco-index-bar-sidebar-active"
              data-index="A"
            >
              <div
                class="arco-index-bar-sidebar-item-wrapper"
              >
                A
              </div>
            </div>
            <div
              class="arco-index-bar-sidebar-item"
              data-index="B"
            >
              <div
                class="arco-index-bar-sidebar-item-wrapper"
              >
                B
              </div>
            </div>
            <div
              class="arco-index-bar-sidebar-item"
              data-index="C"
            >
              <div
                class="arco-index-bar-sidebar-item-wrapper"
              >
                C
              </div>
            </div>
            <div
              class="arco-index-bar-sidebar-item"
              data-index="D"
            >
              <div
                class="arco-index-bar-sidebar-item-wrapper"
              >
                D
              </div>
            </div>
            <div
              class="arco-index-bar-sidebar-item"
              data-index="E"
            >
              <div
                class="arco-index-bar-sidebar-item-wrapper"
              >
                E
              </div>
            </div>
            <div
              class="arco-index-bar-sidebar-item"
              data-index="F"
            >
              <div
                class="arco-index-bar-sidebar-item-wrapper"
              >
                F
              </div>
            </div>
            <div
              class="arco-index-bar-sidebar-item"
              data-index="G"
            >
              <div
                class="arco-index-bar-sidebar-item-wrapper"
              >
                G
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="arco-tab-pane mode-swipe"
      />
    </div>
  </div>
</DocumentFragment>
`;
